odl/discr/discr_ops.py

```diff
@@ -0,0 +1,477 @@
+    endpoints with parameter ``edge_order``.
+
+    The returned array has the same shape as the input array ``f``.
+
+    Parameters
+    ----------
+    f : array-like
+         An N-dimensional array
+    out : array-like or `None`, optional
+         An N-dimensional array
+    axis : `int`, optional
+        The axis along which the partial derivative is evaluated. Default: 0
+    dx : `float`, optional
+        Scalar specifying the sample distances in each dimension ``axis``.
+        Default distance: 1.0
+    edge_order : {1, 2}, optional
+        First-order accurate differences (1) can be used at the boundaries
+        if no zero padding is used. Default edge order: 2
+    zero_padding : `bool`, optional
+        Implicit zero padding. Assumes values outside the domain of ``f`` to be
+        zero. Default: `False`
+
+    Returns
+    -------
+    out : `numpy.ndarray`
+        Returns an N-dimensional array of the same shape as ``f``
+
+    Examples
+    --------
+    >>> f = np.array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9.])
+    >>> finite_diff(f)
+    array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
+    >>> finite_diff(f, axis=0, dx=1.0, edge_order=2, zero_padding=False)
+    array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
+    >>> finite_diff(f, dx=0.5)
+    array([ 2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  2.])
+    >>> finite_diff(f, zero_padding=True)
+    array([ 0.5,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. , -4. ])
+    >>> finite_diff(f, zero_padding=False, edge_order=1)
+    array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
+    >>> out = finite_diff(f)
+    >>> out is finite_diff(f)
+    False
+    >>> out is finite_diff(f, out)
+    True
+    """
+    # TODO: implement forward/backward differences
+    # TODO: implement alternative boundary conditions
+
+    if zero_padding is True and edge_order == 1:
+        raise ValueError("zero padding uses second-order accurate "
+                         "differences at boundaries. First-order accurate "
+                         "edges can only be used without zero padding.")
+    f_data = np.asarray(f)
+    ndim = f_data.ndim
+
+    if not 0 <= axis < ndim:
+        raise IndexError("axis parameter ({0}) exceeds the number of "
+                         "dimensions ({1}).".format(axis, ndim))
+
+    if f_data.shape[axis] < 2:
+        raise ValueError("shape of array too small to calculate a numerical "
+                         "gradient, at least two elements are required.")
+    if out is None:
+        out = np.empty_like(f_data)
+    else:
+        if not out.shape == f.shape:
+            raise ValueError(
+                "shape of `out` array ({0}) does not match the shape of "
+                "input array `f` ({1}).".format(out.shape, f.shape))
+
+    if edge_order not in {1, 2}:
+        raise ValueError("edge order ({0}) not valid".format(edge_order))
+
+    if dx == 0:
```

Comment:
I wouldn't generally disallow it *if* it makes sense. It may be useful, at least in this low-level helper. But that's something which can be done later.